tests/integration/conftest.py

```diff
@@ -1,16 +1,52 @@
 # Copyright 2021 Canonical Ltd.
 # See LICENSE file for licensing details.
+import functools
+import logging
 import shutil
+from collections import defaultdict
+from datetime import datetime
 
 import pytest
 from pytest_operator.plugin import OpsTest
 
+logger = logging.getLogger(__name__)
 
-@pytest.fixture(scope="module")
-async def loki_charm(ops_test: OpsTest):
-    """Loki charm used for integration testing."""
-    charm = await ops_test.build_charm(".")
-    return charm
+
+class Store(defaultdict):
+    def __init__(self):
+        super(Store, self).__init__(Store)
+
+    def __getattr__(self, key):
+        """Override __getattr__ so dot syntax works on keys."""
+        try:
+            return self[key]
+        except KeyError:
+            raise AttributeError(key)
+
+    def __setattr__(self, key, value):
+        """Override __setattr__ so dot syntax works on keys."""
+        self[key] = value
+
+
+store = Store()
+
+
+def timed_memoizer(func):
```

Comment:
It's not required to address this specifically, it's just cutting down the turnaround time for running the test suite by only building the charms once rather than every time. This is, incidentally, part of the on-hold "testing harness", but is useful for testing here rather than adding an additional 1-4m per test file.

Reply:
Thanks for clarifying. I am still inclined to request this be moved into a separate PR if for no reason other than to clarify through the means of this PR precisely those issues that have adversarially impacted the reliability of integration tests in this Loki charm. But if you feel strongly about this I am happy to leave it as is.